### PR TITLE
NLog - ValueSerializer - Faster integer and enum

### LINKED ISSUE #2323: ValueSerializer: use  Convert.GetTypeCode(value)
https://github.com/NLog/NLog/blob/master/src/NLog/MessageTemplates/ValueSerializer.cs

use  Convert.GetTypeCode(value) instead of `as`? 

Needs a benchmark I guess

### PR DESCRIPTION
**Pull request by snakefoot**
Trying to resolve #2323

### CODE REVIEW

File: src/NLog/MessageTemplates/ValueSerializer.cs [hunk 1]
```diff
@@ -59,6 +59,8 @@ private ValueSerializer()
         private const int MaxValueLength = 512 * 1024;
         private const string LiteralFormatSymbol = "l";
 
+        private readonly MruCache<Enum, string> _enumCache = new MruCache<Enum, string>(10000);
```
Comment (304NotModified) [inline src/NLog/MessageTemplates/ValueSerializer.cs]: isn't that a bit big? (10000)

Comment (snakefoot) [reply src/NLog/MessageTemplates/ValueSerializer.cs]: No idea. Just chose the same as for the DefaultJsonSerializer. Maybe there are more than 10000 enums in the world?

**Comment by snakefoot** [comment]
@304NotModified Have cheated a little as Integer.ToString is pretty much the same in all languages. And have also performed the same cheating for Enum.

File: src/NLog/MessageTemplates/ValueSerializer.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/MessageTemplates/ValueSerializer.cs]: Should we change it to 5000 for the both of them ?

**Comment by 304NotModified** [comment]
> @304NotModified Have cheated a little as Integer.ToString is pretty much the same in all languages (when no format-string). And have also performed the same cheating for Enum (when no format string).

No problem if it's document in the source ;)

Otherwise it could get lost in refactoring

**Comment by snakefoot** [comment]
> No problem if it's document in the source ;)

Added a comment about the special optimization.

File: src/NLog/MessageTemplates/ValueSerializer.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/MessageTemplates/ValueSerializer.cs]: was more thinking of 1000. Or even lower. Could not image your using 1000 different enum values in 1 project.

**Comment by 304NotModified** [comment]
do we need to pref test this? Or are you sure it's faster?

**Comment by snakefoot** [comment]
>do we need to pref test this?

The enum printing is 100 pct faster, and the integer printing is 50 pct. faster.

File: src/NLog/MessageTemplates/ValueSerializer.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/MessageTemplates/ValueSerializer.cs]: Now changed it to 1500 (also for the Json-Converter)

**Comment by snakefoot** [comment]
@304NotModified Should be ready for review (again) :)

**Comment by 304NotModified** [comment]
great! Merged!

**Pull request merged**